### PR TITLE
SW-80 cloud config for update

### PR DESCRIPTION
**Pull request by JFSolution**
Changing the update mechanism to look in a json file for infos like the repository, branch, ....
This file will be hostet in the internet and sync at every start of the device or at a tier change

### CODE REVIEW

File: octoprint_mrbeam/software_update_information.py [hunk 1]
```diff
@@ -1,69 +1,178 @@
...
+
+    def look_for_new_config_file(self):
+        """
+        starts a thread to look online for a new config file
+        """
+        th = threading.Thread(target=self._load_update_file_from_cloud)
+        th.setName("MrBeamSoftwareupdateHandler:_load_update_file_from_cloud")
+        th.daemon = True
+        th.start()
+        th.join()
```
Comment (andywerner) [inline octoprint_mrbeam/software_update_information.py]: please remove the join

File: octoprint_mrbeam/software_update_information.py [hunk 2]
```diff
@@ -1,69 +1,178 @@
...
+
+    def _load_update_file_from_cloud(self):
+        """
+        overrides the local update config file if there is a newer one on the server
+        """
+        _logger.debug("load update file")
+        newfile = False
+        try:
+            r = requests.get(
```
Comment (andywerner) [inline octoprint_mrbeam/software_update_information.py]: what happens if server returns an error (400, 500, ...?)

File: octoprint_mrbeam/software_update_information.py [hunk 3]
```diff
@@ -1,69 +1,178 @@
...
+                                update_info["version"],
+                            )
+
+                    except ValueError:
+                        newfile = True
+                        _logger.error(
+                            "there is a wrong configured config local file - override local file with server file"
+                        )
+                    finally:
+                        # override local file
```
Comment (andywerner) [inline octoprint_mrbeam/software_update_information.py]: Please do now re-write the file if it is the same version

File: octoprint_mrbeam/software_update_information.py [hunk 4]
```diff
@@ -1,69 +1,178 @@
...
+        starts a thread to look online for a new config file
+        """
+        th = threading.Thread(target=self._load_update_file_from_cloud)
+        th.setName("MrBeamSoftwareupdateHandler:_load_update_file_from_cloud")
+        th.daemon = True
+        th.start()
+        th.join()
+
+    def _load_update_file_from_cloud(self):
+        """
```
Comment (andywerner) [inline octoprint_mrbeam/software_update_information.py:71]: Because this runs in a separate thread, please add a try/catch all and log the exception at the top most level of the thread.

File: octoprint_mrbeam/static/updates/update_info.json [hunk 5]
```diff
@@ -0,0 +1,123 @@
+{
```
Comment (andywerner) [inline octoprint_mrbeam/static/updates/update_info.json]: this should be removed

Comment (andywerner) [reply octoprint_mrbeam/static/updates/update_info.json]: We need to make sure that the SW always shows the currently installed versions of the SW components - even if we do not have a locally downloaded version and no internet connection.

**Review by andywerner — CHANGES_REQUESTED** [review]
We need to make sure that the SW always shows the currently installed versions of the SW components - even if we do not have a locally downloaded version and no internet connection.

File: octoprint_mrbeam/software_update_information.py (hunk 2 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py]: added exception handling for this case

File: octoprint_mrbeam/software_update_information.py [hunk 6]
```diff
@@ -1,69 +1,178 @@
...
+        self._beamos_date = beamos_date
+
+    def look_for_new_config_file(self):
+        """
+        starts a thread to look online for a new config file
+        """
+        th = threading.Thread(target=self._load_update_file_from_cloud)
+        th.setName("MrBeamSoftwareupdateHandler:_load_update_file_from_cloud")
+        th.daemon = True
+        th.start()
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py]: 1. If you want to use a separate thread, please use the `get_thread` decorator: it automatically sets the name and logs any uncaught exception. If the Only purpose of this function is to create a Thread, then you can just decorate `_load_update_file_from_cloud` with the `@get_thread(daemon=True)` decorator, and you won't need to keep `look_for_new_config_file()`
2. What is the advantage of creating a separate thread here if you are going to join it directly? It looks like this can be done synchronously here.

File: octoprint_mrbeam/software_update_information.py (hunk 4 above)
Comment (amrsoll) [reply octoprint_mrbeam/software_update_information.py:71]: See my previous comment

Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py:71]: ok

File: octoprint_mrbeam/static/updates/update_info.json (hunk 5 above)
Comment (JFSolution) [reply octoprint_mrbeam/static/updates/update_info.json]: ok, it will fallback to mrbeamplugin file if there is no from server or localy

File: octoprint_mrbeam/software_update_information.py [hunk 7]
```diff
@@ -1,69 +1,178 @@
...
+                                "update local file from server - %s -> %s",
+                                serverfile_info["version"],
+                                update_info["version"],
+                            )
+
+                    except ValueError:
+                        newfile = True
+                        _logger.error(
+                            "there is a wrong configured config local file - override local file with server file"
+                        )
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py]: When it comes to un-sanitised content like this JSON info, I would suggest to use `dict.get()`
```python
server_version = serverfile_info.get("version", None)
update_version = update_info.get("version", None)
if server_version is not None:
  ...
```
This way you 
1. don't have to try/except the `valueError` if the version is not available 
2. Can decide what to use when `version` is not available, either from the server or from the update_info.

File: octoprint_mrbeam/software_update_information.py [hunk 8]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+    @param _softwareupdate_handler: the handler class to look for a new config file online
+    """
+    try:
+        with open(join(plugin._settings.getBaseFolder("base"), SW_UPDATE_FILE)) as f:
+            update_info = json.load(f)
+    except IOError as e:
+        _softwareupdate_handler.look_for_new_config_file()
+        _logger.error(
+            "could not find file %s - error: %s",
+            join(plugin._settings.getBaseFolder("base"), SW_UPDATE_FILE),
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:324]: Please put this path into a function if you need to create it dynamically. You are using the same line too many times. (Rule of thumb: do not repeat yourself more than 3 times)

File: octoprint_mrbeam/software_update_information.py [hunk 9]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+            e,
+        )
+    except ValueError as e:
+        _softwareupdate_handler.look_for_new_config_file()
+        _logger.error(
+            "there is a wrong configured config file local - try to load from server"
+        )
+        _logger.debug("error %s - %s", e, f)
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:332]: `f` might not be available here.
Better to log the path instead of the file object as well.

File: octoprint_mrbeam/software_update_information.py [hunk 10]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
     else:
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:333]: This `else` statement goes aaaall the way to the end of the function. Please return when the exceptions happen instead of indenting 130 lines.

File: octoprint_mrbeam/software_update_information.py [hunk 11]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
     )
 
-
-def _set_info_mrb_hw_info(plugin, tier, beamos_date):
-    if beamos_date > BEAMOS_LEGACY_DATE:
-        pip_command = "sudo /usr/local/iobeam/venv/bin/pip"
+    # plugin._settings.global_set_boolean(op_swu_keys + ["prerelease"], False)
+
+
+def _set_info_from_file(plugin, tier, beamos_date, _softwareupdate_handler):
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:294]: This function is too big. I would recommend to split it up into multiple functions with good descriptive names of what they are doing.

File: octoprint_mrbeam/software_update_information.py [hunk 12]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+                            "displayVersion": current_version,
+                        }
+                    )
+            except:
+                _logger.error("Error in module %s %s", module_id, sys.exc_info()[0])
+        _logger.debug(sw_update_config)
+
+
+def _set_update_config_from_dict(update_config, dict):
+    update_config.update(dict)
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:465]: 1. This function on the other hand is too small to make any sense.
2. Did you mean to make a shallow merge? Otherwise, please use the `octoprint.util.dict_merge()` function to make a deep merge of the tree structure

File: octoprint_mrbeam/software_update_information.py [hunk 13]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+
+
+def _get_display_name(plugin, name):
+    return name
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:469]: Same here, Did you mean to add more content to this function for this PR?
If yes, then please set the PR as draft until the changes are reflected
If not, then please do not add extra functions for potential future code, because then it gets lost.

File: octoprint_mrbeam/software_update_information.py [hunk 14]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+    tierversion = get_tier_by_id(tier)
+    if tierversion in config:
+        module_tier = config[tierversion]
+        _set_update_config_from_dict(
+            config, module_tier
+        )  # set tier config from default settings
+
+    if "beamos_date" in config:
+        beamos_date_config = config["beamos_date"]
+        for date, beamos_config in beamos_date_config.items():
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:267]: Tip: you can shorten this to
```python
for date, beamos_config in config.get("beamos_date" ,{}).items():
    ...
```

File: octoprint_mrbeam/software_update_information.py [hunk 15]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+                    if tierversion in moduleconfig:
+                        module_tier = moduleconfig[tierversion]
+                        _set_update_config_from_dict(
+                            moduleconfig, module_tier
+                        )  # set tier config from default settings
+                    if tierversion in module:
+                        module_tier = module[tierversion]
+                        _set_update_config_from_dict(
+                            moduleconfig, module_tier
+                        )  # override tier config from tiers set in config_file
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:373]: the Black formatter does not like comments at the end of lines

Better to have them in a separate line (the previous line) so they don't create these ugly parenthesis-newline patterns

File: octoprint_mrbeam/software_update_information.py [hunk 16]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+
+                    _set_update_config_from_dict(
+                        moduleconfig, module
+                    )  # set default config from file for module
+
+                    # have to be after the default config from file
+                    if "beamos_date" in module:
+                        beamos_date_config = module["beamos_date"]
+                        for date, beamos_config in beamos_date_config.items():
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:382]: My previous tip also works here :)

File: octoprint_mrbeam/software_update_information.py [hunk 17]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+                            _logger.debug(
+                                "date compare %s >= %s -> %s",
+                                beamos_date,
+                                datetime.strptime(date, "%Y-%m-%d").date(),
+                                beamos_config,
+                            )
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:388]: Please do not log if conditions, that's too much...

File: octoprint_mrbeam/software_update_information.py [hunk 18]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+                                if tierversion in beamos_config:
+                                    beamos_config_module_tier = beamos_config[
+                                        tierversion
+                                    ]
+                                    _set_update_config_from_dict(
+                                        beamos_config, beamos_config_module_tier
+                                    )  # override tier config from tiers set in config_file
+                                _set_update_config_from_dict(
+                                    moduleconfig, beamos_config
+                                )
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:402]: This is where a deep merge using `dict_merge` is nice; I think it can replace this extra `if` statement

File: octoprint_mrbeam/software_update_information.py [hunk 19]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+                        moduleconfig["pip_command"] = GLOBAL_PIP_COMMAND
+                    if "pip_command" in moduleconfig:
+                        pip_command = moduleconfig["pip_command"]
+                        # if global_pip_command is set module is installed outside of our virtualenv therefor we can't use default pip command.
+                        # /usr/local/lib/python2.7/dist-packages must be writable for pi user otherwise OctoPrint won't accept this as a valid pip command
+                        # pip_command = GLOBAL_PIP_COMMAND
+                        package_name = (
+                            module["package_name"]
+                            if "package_name" in module
+                            else module_id
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:430]:
```python
package_name = module.get("package_name", module_id)
```

File: octoprint_mrbeam/software_update_information.py [hunk 20]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
+                        module,
+                        moduleconfig,
+                        "pip_command" not in moduleconfig,
+                    )
+                    if (
+                        "global_pip_command" in module
+                        and "pip_command" not in moduleconfig
+                    ):
+                        moduleconfig["pip_command"] = GLOBAL_PIP_COMMAND
+                    if "pip_command" in moduleconfig:
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:422]: This is more for readability, but I guess this `if` could be an `elif` ? It helps with understanding what we are doing.

File: octoprint_mrbeam/software_update_information.py [hunk 21]
```diff
@@ -100,231 +215,244 @@ def switch_software_channel(plugin, channel):
...
-        branch=branch,
-        branch_default=branch_default,
-        restart=restart,
-        **kwargs
-    )
-    return {module_id: update_info}
+        try:
+            _set_octoprint_config(
+                plugin, tier, update_info.pop("octoprint"), beamos_date
+            )
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py]: Is there a reason why you want to `.pop()` those values instead of `.get()` ? Couldn't it lead to unpredicted missing info later on?

File: octoprint_mrbeam/software_update_information.py [hunk 22]
```diff
@@ -1,69 +1,178 @@
...
 GLOBAL_PIP_BIN = "/usr/local/bin/pip"
 GLOBAL_PIP_COMMAND = (
     "sudo {}".format(GLOBAL_PIP_BIN) if os.path.isfile(GLOBAL_PIP_BIN) else None
 )
-# GLOBAL_PIP_COMMAND = "sudo {} -m pip".format(GLOBAL_PY_BIN) if os.path.isfile(GLOBAL_PY_BIN) else None #  --disable-pip-version-check
-# VENV_PIP_COMMAND = ("%s -m pip --disable-pip-version-check" % VENV_PY_BIN).split(' ') if os.path.isfile(VENV_PY_BIN) else None
+
 BEAMOS_LEGACY_DATE = date(2018, 1, 12)
 
+_softwareupdate_handler = None
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:45]: Why are we using this as a global?

File: octoprint_mrbeam/software_update_information.py [hunk 23]
```diff
@@ -1,69 +1,178 @@
...
 
+def get_modules():
+    return sw_update_config
+
+
+class MrBeamSoftwareupdateHandler:
+    def __init__(self, plugin, tier, beamos_date):
+        self._plugin = plugin
+        self._tier = tier
+        self._beamos_date = beamos_date
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:56]: Why is this class necessary?
It does not seem like you are using it's attributes, and `self` is only used for referencing `self._plugin`, so to me it looks redundant. I would suggest to keep this on a functional approach instead of being Object Oriented.

File: octoprint_mrbeam/software_update_information.py [hunk 24]
```diff
@@ -1,69 +1,178 @@
...
+SW_UPDATE_CLOUD_PATH = "https://mr-beam.org/beamos/config/sw-update-conf"
+
 # add to the display name to modules that should be shown at the top of the list
 SORT_UP_PREFIX = " "
 
-
 _logger = mrb_logger("octoprint.plugins.mrbeam.software_update_information")
 
+sw_update_config = dict()
+
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:34]: Please do NOT use mutable globals as much as possible.

File: octoprint_mrbeam/software_update_information.py [hunk 25]
```diff
@@ -1,69 +1,178 @@
...
-                _set_info_mrb_hw_info(plugin, tier, beamos_date),
-                # _config_octoprint(plugin, tier),
-            ],
-        )
-    )
+    _logger.info("SoftwareUpdate using tier: %s %s", tier, beamos_date)
+    global _softwareupdate_handler
+    if _softwareupdate_handler is None:
+        _softwareupdate_handler = MrBeamSoftwareupdateHandler(plugin, tier, beamos_date)
+        _softwareupdate_handler.look_for_new_config_file()
```
Comment (amrsoll) [inline octoprint_mrbeam/software_update_information.py:191]: If you decide to remove the `MrBeamSoftwareupdateHandler` class, then you can just pass the plugin directly to `look_for_new_config_file(plugin)`.

**Review by amrsoll — CHANGES_REQUESTED** [review]
Hey, I would like to have some of my concerns addressed if possible.

I had refactored this part to be more function-based with immutable variables. I do not think this is the right approach to solve the problem.

File: octoprint_mrbeam/software_update_information.py (hunk 6 above)
Comment (amrsoll) [reply octoprint_mrbeam/software_update_information.py]: N.B. when a function is decorated with `get_thread`, it will return a `Thread` object which you can decide to `.join()` later on

Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py]: the join was not intended, the thread was requested from andy so the eventhandler will not be blocked by this call

File: octoprint_mrbeam/software_update_information.py (hunk 7 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py]: ok this looks neat, I will try this

File: octoprint_mrbeam/software_update_information.py (hunk 8 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py:324]: ok will refactor this

File: octoprint_mrbeam/software_update_information.py (hunk 9 above)
Comment (Josef-MrBeam) [reply octoprint_mrbeam/software_update_information.py:300]: ok will change

File: octoprint_mrbeam/software_update_information.py (hunk 21 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py]: I pop it so it wont be handeld by the for loop that processes all the module infos

File: octoprint_mrbeam/software_update_information.py (hunk 22 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py:45]: To make sure that there are not multiple calls that run at the same time

File: octoprint_mrbeam/software_update_information.py (hunk 21 above)
Comment (amrsoll) [reply octoprint_mrbeam/software_update_information.py]: In that case I would say that they shouldn't be linked to the same node if they are not going to be treated the same; so we don't have to do risky things like this.

File: octoprint_mrbeam/software_update_information.py (hunk 22 above)
Comment (amrsoll) [reply octoprint_mrbeam/software_update_information.py:45]: Then I would suggest to use a `Lock` / `Rlock` instead :)

File: octoprint_mrbeam/software_update_information.py (hunk 23 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py:56]: hmm, I can't remember why I choose to do it Object oriented, you are right I only use the self._plugin
What is the downside to do it this way?

File: octoprint_mrbeam/software_update_information.py (hunk 24 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py:34]: I looked it up I used it as it was before your buster changes, I can change this like you did for the buster image

File: octoprint_mrbeam/software_update_information.py (hunk 3 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py]: ok, changed

File: octoprint_mrbeam/software_update_information.py (hunk 6 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py]: shouldn't it be without the daemon, this could cause file corruption when it gets killed while writing

File: octoprint_mrbeam/software_update_information.py (hunk 9 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py:332]: the idea was to log what is the problem in the file and it should be available, because its just a valueerror of the json.load

File: octoprint_mrbeam/software_update_information.py (hunk 21 above)
Comment (JFSolution) [reply octoprint_mrbeam/software_update_information.py]: Ok I can put it in a separate node, but why it is risky to pop it? I know that I don't need this data later

**Comment by JFSolution** [comment]
new PR https://github.com/mrbeam/MrBeamPlugin/pull/1389
